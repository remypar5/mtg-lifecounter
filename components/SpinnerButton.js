import React from 'react';
import { StyleSheet, TouchableHighlight, Text } from 'react-native';

export default class SpinnerButton extends React.Component {
    constructor(props) {
        super(props);

        this._onPress = props.onPress.bind(this);
        if (props.onLongPress) {
            this._onLongPress = props.onLongPress.bind(this);
        }
    }
    
    render() {
        const props = this.props;
        
        return (
            <TouchableHighlight style={ styles.hitbox } 
                onPress={ this._onPress } onLongPress={ this._onLongPress }>
                <Text style={ styles.button }>{ props.content }</Text>
            </TouchableHighlight>
        );
    }
}

const styles = StyleSheet.create({
    hitbox: {
        width: '40%',
    },
    button: {
        fontSize: 20,
        height: '100%',
        color: 'white',
        fontWeight: 'bold',
        textAlign: 'center',
        textAlignVertical: 'center',
    }
});