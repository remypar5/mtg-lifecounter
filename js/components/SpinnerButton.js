import React from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, TouchableWithoutFeedback, Text, View } from 'react-native';

import { FONT_FAMILY, COLOR_FOREGROUND } from '../utils/constants';

const styles = StyleSheet.create({
    buttonContainer: {
        width: '20%',
    },
    button: {
        fontSize: 32,
        fontFamily: FONT_FAMILY,
        width: '100%',
        height: '100%',
        color: COLOR_FOREGROUND,
        textAlign: 'center',
        textAlignVertical: 'center',
    },
});

export default class SpinnerButton extends React.Component {
    constructor(props) {
        super(props);

        this.preventOnPress = false;
        this.onPress = this.onPress.bind(this);

        if (props.onHold) {
            this.onPressIn = this.onPressIn.bind(this);
            this.onPressOut = this.onPressOut.bind(this);
        }
    }

    onPress(event) {
        if (this.preventOnPress) {
            return;
        }
        this.props.onPress(event);
        this.preventOnPress = false;
    }

    onPressIn() {
        this.onHoldInterval = setInterval(() => {
            this.preventOnPress = true;
            this.props.onHold();
        }, 650);
    }

    onPressOut() {
        clearInterval(this.onHoldInterval);
        setTimeout(() => {
            this.preventOnPress = false;
        }, 100);
    }

    render() {
        const { content } = this.props;
        const { button, buttonContainer } = styles;

        return (
            <TouchableWithoutFeedback
                onPress={this.onPress}
                onPressIn={this.onPressIn}
                onPressOut={this.onPressOut}
            >
                <View style={[buttonContainer]}>
                    <Text style={button}>{content}</Text>
                </View>
            </TouchableWithoutFeedback>
        );
    }
}

SpinnerButton.propTypes = {
    content: PropTypes.string.isRequired,
    onPress: PropTypes.func.isRequired,
    onHold: PropTypes.func,
};

SpinnerButton.defaultProps = {
    onHold: null,
};
