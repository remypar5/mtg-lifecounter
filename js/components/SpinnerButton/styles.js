import { StyleSheet } from 'react-native';

import { FONT_FAMILY, COLOR_FOREGROUND } from '../../utils/constants';

const styles = StyleSheet.create({
    buttonContainer: {
        width: '20%',
    },
    button: {
        fontSize: 32,
        fontFamily: FONT_FAMILY,
        width: '100%',
        height: '100%',
        color: COLOR_FOREGROUND,
        textAlign: 'center',
        textAlignVertical: 'center',
    },
});

export default styles;
