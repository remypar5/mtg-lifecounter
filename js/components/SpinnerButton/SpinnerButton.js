import React from 'react';
import PropTypes from 'prop-types';
import { TouchableWithoutFeedback, Text, View } from 'react-native';

import styles from './styles';

export default class SpinnerButton extends React.Component {
    constructor(props) {
        super(props);

        this.preventOnPress = false;
        this.onPress = this.onPress.bind(this);

        if (props.onHold) {
            this.onPressIn = this.onPressIn.bind(this);
            this.onPressOut = this.onPressOut.bind(this);
        }
    }

    onPress(event) {
        if (this.preventOnPress) {
            return;
        }
        this.props.onPress(event);
        this.preventOnPress = false;
    }

    onPressIn() {
        this.onHoldInterval = setInterval(() => {
            this.preventOnPress = true;
            this.props.onHold();
        }, 650);
    }

    onPressOut() {
        clearInterval(this.onHoldInterval);
        setTimeout(() => {
            this.preventOnPress = false;
        }, 100);
    }

    render() {
        const { content } = this.props;
        const { button, buttonContainer } = styles;
        const hitSlop = {
            top: 100,
            bottom: 100,
            left: 100,
            right: 100,
        };

        return (
            <TouchableWithoutFeedback
                hitSlop={hitSlop}
                onPress={this.onPress}
                onPressIn={this.onPressIn}
                onPressOut={this.onPressOut}
            >
                <View style={[buttonContainer]}>
                    <Text style={button}>{content}</Text>
                </View>
            </TouchableWithoutFeedback>
        );
    }
}

SpinnerButton.propTypes = {
    content: PropTypes.string.isRequired,
    onPress: PropTypes.func.isRequired,
    onHold: PropTypes.func,
};

SpinnerButton.defaultProps = {
    onHold: null,
};
