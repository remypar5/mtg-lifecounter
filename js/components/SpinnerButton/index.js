import SpinnerButton from './SpinnerButton';
import styles from './styles';

export default SpinnerButton;
export {
    styles,
};
