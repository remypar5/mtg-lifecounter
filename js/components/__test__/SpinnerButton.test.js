import React from 'react';
import { shallow } from 'enzyme';

import SpinnerButton from '../SpinnerButton';

describe('<SpinnerButton />', () => {

    let component;
    let onPressSpy = jest.fn();
    let onHoldSpy = jest.fn();

    beforeEach(() => {
        component = shallow(
            <SpinnerButton
                content="-"
                onPress={ onPressSpy } />
            );

        onPressSpy.mockReset();
    });

    it('renders without crashing', () => {
        expect(component).toMatchSnapshot();
    });

    it('should call the onPress callback when it is pressed', () => {
        expect(onPressSpy).not.toHaveBeenCalled();

        component.simulate('press');
        expect(onPressSpy).toHaveBeenCalled();
    });

    describe('holding it down', () => {
        beforeEach(() => {
            jest.useFakeTimers();

            component = shallow(
                <SpinnerButton
                    onPress={ onPressSpy }
                    onHold={ onHoldSpy } />
            );

            onPressSpy.mockReset();
            onHoldSpy.mockReset();
        });

        it('should call only the onHold callback when it is held for 900ms', () => {
            expect(onPressSpy).not.toHaveBeenCalled();
            expect(onHoldSpy).not.toHaveBeenCalled();

            component.simulate('pressIn');
            expect(onHoldSpy).not.toHaveBeenCalled();

            jest.runOnlyPendingTimers();
            component.simulate('pressOut');

            expect(onHoldSpy).toHaveBeenCalled();
            expect(onPressSpy).not.toHaveBeenCalled();
        });

        it('should call only the onHold callback once every 900ms it is held', () => {
            expect(onPressSpy).not.toHaveBeenCalled();
            expect(onHoldSpy).not.toHaveBeenCalled();

            component.simulate('pressIn');
            expect(onHoldSpy).not.toHaveBeenCalled();

            jest.runOnlyPendingTimers();
            jest.runOnlyPendingTimers();
            component.simulate('pressOut');

            expect(onHoldSpy).toHaveBeenCalledTimes(2);
            expect(onPressSpy).not.toHaveBeenCalled();
        });
    });
});
