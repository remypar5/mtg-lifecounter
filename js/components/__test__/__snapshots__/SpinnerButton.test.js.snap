// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpinnerButton /> renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SpinnerButton
    content="-"
    onPress={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <View
        style={
          Array [
            Object {
              "width": "20%",
            },
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#b2b2b0",
              "fontFamily": "cloisterblack",
              "fontSize": 32,
              "height": "100%",
              "textAlign": "center",
              "textAlignVertical": "center",
              "width": "100%",
            }
          }
        >
          -
        </Text>
      </View>,
      "onPress": [Function],
      "onPressIn": [Function],
      "onPressOut": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#b2b2b0",
              "fontFamily": "cloisterblack",
              "fontSize": 32,
              "height": "100%",
              "textAlign": "center",
              "textAlignVertical": "center",
              "width": "100%",
            }
          }
        >
          -
        </Text>,
        "style": Array [
          Object {
            "width": "20%",
          },
        ],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "accessible": true,
          "allowFontScaling": true,
          "children": "-",
          "disabled": false,
          "ellipsizeMode": "tail",
          "style": Object {
            "color": "#b2b2b0",
            "fontFamily": "cloisterblack",
            "fontSize": 32,
            "height": "100%",
            "textAlign": "center",
            "textAlignVertical": "center",
            "width": "100%",
          },
        },
        "ref": null,
        "rendered": "-",
        "type": [Function],
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <View
          style={
            Array [
              Object {
                "width": "20%",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#b2b2b0",
                "fontFamily": "cloisterblack",
                "fontSize": 32,
                "height": "100%",
                "textAlign": "center",
                "textAlignVertical": "center",
                "width": "100%",
              }
            }
          >
            -
          </Text>
        </View>,
        "onPress": [Function],
        "onPressIn": [Function],
        "onPressOut": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#b2b2b0",
                "fontFamily": "cloisterblack",
                "fontSize": 32,
                "height": "100%",
                "textAlign": "center",
                "textAlignVertical": "center",
                "width": "100%",
              }
            }
          >
            -
          </Text>,
          "style": Array [
            Object {
              "width": "20%",
            },
          ],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "accessible": true,
            "allowFontScaling": true,
            "children": "-",
            "disabled": false,
            "ellipsizeMode": "tail",
            "style": Object {
              "color": "#b2b2b0",
              "fontFamily": "cloisterblack",
              "fontSize": 32,
              "height": "100%",
              "textAlign": "center",
              "textAlignVertical": "center",
              "width": "100%",
            },
          },
          "ref": null,
          "rendered": "-",
          "type": [Function],
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
